### PR TITLE
Add ExpandablePanel module

### PR DESCRIPTION
**Pull request by diogo-mendes**
The relative complexity of the ExpandablePanel component is due mainly to the sliding animation. Although it is not buttery smooth, it is functional. 

I would ask you to check out the branch and try it out for yourselves in the storybook, so that maybe you can come up with a suggestion to improve the animation. One thing that can probably be improved is the setTimeout used to set the bodyHeight (`ExpandablePanel.tsx` line 128).

### ExpandablePanel

The `ExpandablePanel` is very self-explanatory - it is a content panel that expands/collapses when clicked on.
Its expanded/collapsed state can either be controlled by the component internally or by a parent component.

### ExpandablePanels

`ExpandablePanel`s can be grouped in an `ExpandablePanels` component. Multiple `ExpandablePanel`s can be opened simultaneously by default.
However, if the `allowMultipleExpandedPanels` prop is passed as `false`, `ExpandablePanels` can mimic the behaviour of an Accordion, where expanding
one panel will collapse the other.
This component also passes other props to its children, e.g. `onClick` or `disabled`.

### CODE REVIEW

File: modules/expandable-panel/ExpandablePanel.story.mdx [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+      <td>disabled (optional)</td>
+      <td>boolean</td>
+      <td>
+        Flag indicating whether user can interact with any of the
+        ExpandablePanel components
+      </td>
+    </tr>
+    <tr>
+      <td>initialExpandedPanelIndex (optional)</td>
+      <td>number</td>
```
Comment (acaldas) [inline modules/expandable-panel/ExpandablePanel.story.mdx]: We could change this to be `number | number[]` to allow setting multiple initial expanded panels, this way we don't have a situation where this prop is "invalid". What do you think?

File: modules/expandable-panel/ExpandablePanel.tsx [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+    // allow the animation to finish before removing the body from the DOM
+    setTimeout(() => {
+      setIsCollapsing(false);
+    }, SLIDE_ANIMATION_DURATION_MS);
+  }, []);
+
+  // set body height manually to be able to have the sliding animation
+  useEffect(() => {
+    if (isExpanded && bodyHeight && bodyHeight > 0 && currentBodyHeight === 0) {
+      setTimeout(() => {
```
Comment (acaldas) [inline modules/expandable-panel/ExpandablePanel.tsx]: We should clear the timeout when an effect is triggered too fast to avoid leaks and side effects.

```
useEffect(() => {
const timeout = setTimeout(...);
return () => clearTimeout(timeout);
}

**Comment by acaldas** [comment]
Did you try animating scaleY from 0 to 1 instead of height? I have used this on the past and it worked okish.

File: modules/base/aphroditeAdapter.ts [hunk 3]
```diff
@@ -0,0 +1,193 @@
+/* eslint-disable class-methods-use-this */
```
Comment (migue1coel6odev) [inline modules/base/aphroditeAdapter.ts]: should we create an issue to update aphrodite?

File: modules/expandable-panel/ExpandablePanel.tsx [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+  ThemeStyleSheetFactory,
+  Theme
+} from "@diana-ui/types";
+import { useWindowSize } from "@diana-ui/hooks";
+import { Icon } from "@diana-ui/icon";
+import { TextHighlight } from "@diana-ui/typography";
+
+const SLIDE_ANIMATION_DURATION_MS = 200;
+
+// @ts-ignore
```
Comment (migue1coel6odev) [inline modules/expandable-panel/ExpandablePanel.tsx]: why is this needed ?

File: modules/expandable-panel/ExpandablePanel.tsx [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+  return (
+    <div
+      className={cx(className, styles.panel, ...stateClasses)}
+      style={{ maxHeight: canAnimate ? "initial" : `${headerHeight}px` }}
+      onClick={handleClick}
+    >
+      {renderHeader()}
+      {(isExpanded || isCollapsing) && (
+        <div
+          ref={bodyWrapperRef}
```
Comment (migue1coel6odev) [inline modules/expandable-panel/ExpandablePanel.tsx]: This ref isn't being used from what I can see right ? can we delete it? 

And is it possible to concat this div and the one below into one ? or this messes up with the animations ?

File: modules/expandable-panel/ExpandablePanel.story.mdx (hunk 1 above)
Comment (diogo-mendes) [reply modules/expandable-panel/ExpandablePanel.story.mdx]: I also considered that.
My reasoning was that this prop would be used if the `allowMultipleExpandedPanels` prop would be used as well. So if ÈxpandablePanels` were to act as an accordion, `initialExpandedPanelIndex ` is passed. Otherwise, the `initialExpanded` prop can be passed to each of the child components.

```
initialExpanded: allowMultipleExpandedPanels
  ? initialExpandedPanelIndex === index || child.props.initialExpanded
  : initialExpandedPanelIndex === index
```

That being said, do you still think your suggestion should still be implemented? If so, we would also have to check if it is an accordion and only expand the first index of the array?

File: modules/expandable-panel/ExpandablePanel.tsx (hunk 2 above)
Comment (diogo-mendes) [reply modules/expandable-panel/ExpandablePanel.tsx]: Absolutely 👍

File: modules/expandable-panel/ExpandablePanel.tsx [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+import { useWindowSize } from "@diana-ui/hooks";
+import { Icon } from "@diana-ui/icon";
+import { TextHighlight } from "@diana-ui/typography";
+
+const SLIDE_ANIMATION_DURATION_MS = 200;
+
+// @ts-ignore
+export interface IProps extends StandardProps<"div"> {
+  disabled?: boolean;
+  expanded?: boolean;
```
Comment (migue1coel6odev) [inline modules/expandable-panel/ExpandablePanel.tsx]: not sure why we need this prop here since the state is managed internally. am I missing something ?

File: modules/expandable-panel/ExpandablePanel.story.mdx (hunk 1 above)
Comment (acaldas) [reply modules/expandable-panel/ExpandablePanel.story.mdx]: I'd say that if we implement this then the user is responsible for correctly using the array prop

File: modules/expandable-panel/ExpandablePanel.tsx (hunk 6 above)
Comment (migue1coel6odev) [reply modules/expandable-panel/ExpandablePanel.tsx]: if this is for the multiple panels could we have a `useImperativeHandle` for each panel?

File: modules/expandable-panel/ExpandablePanel.tsx [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+  const canAnimate = isBodyHeightReady && (isExpanded || isCollapsing);
+  const finalBodyHeight = canAnimate ? currentBodyHeight : "100%";
+  const bodyWrapperStyles = {
+    height: finalBodyHeight,
+    maxHeight: finalBodyHeight === 0 && !isCollapsing ? 0 : undefined
+  };
+
+  return (
+    <div
+      className={cx(className, styles.panel, ...stateClasses)}
```
Comment (migue1coel6odev) [inline modules/expandable-panel/ExpandablePanel.tsx]: instead of having this `...stateClasses` throught the different pieces of components. we could have a selector that does thit for us. Something like: 
```javascript
const stylesheet: ThemeStyleSheetFactory = (theme: Theme) => ({
  panel: {
    overflow: "hidden",
    "@selectors": {
         "&.disabled .header" {
                ...apply styles for header here
          }
   }
  }
});
```

File: modules/expandable-panel/ExpandablePanel.tsx (hunk 4 above)
Comment (diogo-mendes) [reply modules/expandable-panel/ExpandablePanel.tsx]: Since the standard div props are being extended, I can't override onClick with a different signature without TS complaining 😕

File: modules/expandable-panel/ExpandablePanel.tsx (hunk 5 above)
Comment (diogo-mendes) [reply modules/expandable-panel/ExpandablePanel.tsx]: Yes, you're right! I'll remove.

I purposefully added the wrapper div for the animation's sake, yes. In a nutshell, paddings were messing up the height calculation and animations, so this way I keep the body intact (height and padding and all), extract the correct height value for the animation from it, and animate the wrapper div

File: modules/expandable-panel/ExpandablePanel.tsx (hunk 6 above)
Comment (diogo-mendes) [reply modules/expandable-panel/ExpandablePanel.tsx]: `expanded` is used by `ExpandableParents` indeed, and could be used by any other parent component that sees fit to manually control the panel's state.

What is the benefit of using `useImperativeHandle` over a prop? As I understand it using it is ill-advised save for a few exceptions. (Maybe this is one of them?)

Comment (migue1coel6odev) [reply modules/expandable-panel/ExpandablePanel.tsx]: I just suggest that so we don’t need to manage two states (here and the parent component), and it makes a little more confusing on the component itself. Nevertheless its ok for me to keep as you have :)

File: modules/base/aphroditeAdapter.ts (hunk 3 above)
Comment (diogo-mendes) [reply modules/base/aphroditeAdapter.ts]: At least this version of aesthetic I don't think they (he) will care, since he's working on another version already and this is considered legacy.

Maybe when we allocate the effort of migrating to the new version there will be a better way, although I don't know if that version is ready already

File: modules/expandable-panel/ExpandablePanel.tsx (hunk 7 above)
Comment (diogo-mendes) [reply modules/expandable-panel/ExpandablePanel.tsx]: I think the pattern we have adhered to is to apply the modifier class to the necessary elements instead of your suggestion. For consistency's sake, I will leave as is and we can discuss as a team whether we want to change this pattern.

File: modules/expandable-panel/ExpandablePanel.story.mdx (hunk 1 above)
Comment (diogo-mendes) [reply modules/expandable-panel/ExpandablePanel.story.mdx]: For simplicity sake I suggest leaving it as is. If you feel strongly that we should add this change, I can add it.

File: modules/base/package.json [hunk 8]
```diff
@@ -28,6 +28,7 @@
     "aesthetic": "^5.1.0",
     "aesthetic-adapter-aphrodite": "^5.1.0",
```
Comment (nunoliveira12) [inline modules/base/package.json]: should we then remove this one?

Comment (nunoliveira12) [reply modules/base/package.json]: but we should keep aware of this for future updates, mainly when we do update aesthtetic

Comment (diogo-mendes) [reply modules/base/package.json]: Ok, will do!
